### PR TITLE
windows compatibility

### PR DESCRIPTION
**Pull request by lordvlad**
Chown does not exist on widows. This PR uses nodejs's native fs.chmodSync instead

### CODE REVIEW

File: src/bin/link_in_test_app.ts [hunk 1]
```diff
@@ -15,7 +15,8 @@ fs.writeFileSync(
                 return {
                     ...packageJsonParsed,
                     "main": packageJsonParsed["main"].replace(/^dist\//, ""),
-                    "types": packageJsonParsed["types"].replace(/^dist\//, "")
+                    "types": packageJsonParsed["types"].replace(/^dist\//, ""),
+                    "bin": Object.fromEntries(Object.entries<string>(packageJsonParsed["bin"]).map(([k, v]) => [k, v.replace(/^dist\//, "")]))
```
Comment (lordvlad) [inline src/bin/link_in_test_app.ts:19]: not sure why *nix did not break without this, but on windows it resulted in failures when trying to call `yarn link`

File: src/bin/tools/grant-exec-perms.ts [hunk 2]
```diff
@@ -1,10 +1,11 @@
...
-import * as child_process from "child_process";
-import * as fs from "fs";
+import { chmodSync, statSync, constants } from "fs";
 
-Object.entries<string>(JSON.parse(fs.readFileSync(pathJoin(getProjectRoot(), "package.json")).toString("utf8"))["bin"]).forEach(([, scriptPath]) =>
-    child_process.execSync(`chmod +x ${scriptPath}`, {
-        "cwd": getProjectRoot()
-    })
-);
+import(pathJoin(getProjectRoot(), "package.json")).then(({ bin }) => {
```
Comment (lordvlad) [inline src/bin/tools/grant-exec-perms.ts:5]: you can import json files just like modules, no need to battle with fs.read or json.parse

Comment (garronej) [reply src/bin/tools/grant-exec-perms.ts:5]: All right,  
it's just that it's more conveignent to read synchronously.  

But I don't mind you change ⬆️

File: src/bin/link_in_test_app.ts (hunk 1 above)
Comment (garronej) [reply src/bin/link_in_test_app.ts:19]: Oh very nice thanks!

File: src/bin/tools/downloadAndUnzip.ts [hunk 3]
```diff
@@ -1,80 +1,289 @@
...
+ *
+ * If the target file exists and is out of date, it will be overwritten.
+ * If the target file exists and there is no etag file, the target file will
+ * be overwritten.
+ *
+ * @param url download url
+ * @param dir target directory
+ * @returns promise for the full path of the downloaded file
+ */
+async function download(url: string, dir: string): Promise<string> {
```
Comment (lordvlad) [inline src/bin/tools/downloadAndUnzip.ts:53]: It _might_ seem unorthodox to implement this from scratch, but really, introducing another library was not worth it IMHO

File: src/bin/tools/downloadAndUnzip.ts [hunk 4]
```diff
@@ -1,80 +1,289 @@
...
+    }
 
-        fs.rmSync(extractDirPath, { "recursive": true, "force": true });
+    return success;
+}
 
-        fs.mkdirSync(extractDirPath);
+/**
+ *
+ * @param file fileto read
```
Comment (lordvlad) [inline src/bin/tools/downloadAndUnzip.ts:175]:
```suggestion
 * @param file file to read
```

File: src/bin/tools/downloadAndUnzip.ts [hunk 5]
```diff
@@ -1,80 +1,289 @@
...
+    pathOfDirToExtractInArchive,
+    cacheDirPath
+}: {
+    isSilent: boolean;
+    url: string;
+    destDirPath: string;
+    pathOfDirToExtractInArchive?: string;
+    cacheDirPath: string;
+}) {
+    const downloadHash = hash(JSON.stringify({ url, pathOfDirToExtractInArchive })).substring(0, 15);
```
Comment (lordvlad) [inline src/bin/tools/downloadAndUnzip.ts:278]: This is now a bit different to your previous cachin strategy, let me know what you think of it

File: src/bin/tools/grant-exec-perms.ts [hunk 6]
```diff
@@ -1,10 +1,17 @@
...
-Object.entries<string>(JSON.parse(fs.readFileSync(pathJoin(getProjectRoot(), "package.json")).toString("utf8"))["bin"]).forEach(([, scriptPath]) =>
-    child_process.execSync(`chmod +x ${scriptPath}`, {
-        "cwd": getProjectRoot()
-    })
-);
+async () => {
+    var { bin } = await import(pathJoin(getProjectRoot(), "package.json"));
+
+    var promises = Object.values<string>(bin).map(async scriptPath => {
+        const fullPath = pathJoin(getProjectRoot(), scriptPath);
```
Comment (lordvlad) [inline src/bin/tools/grant-exec-perms.ts:10]: a bit more synchronous, eh? :)

File: src/bin/tools/downloadAndUnzip.ts (hunk 3 above)
Comment (garronej) [reply src/bin/tools/downloadAndUnzip.ts:53]: Agreed, mutch apreciated!

File: src/bin/tools/downloadAndUnzip.ts (hunk 5 above)
Comment (garronej) [reply src/bin/tools/downloadAndUnzip.ts:278]: LGTM in surface.  
I didn't test if it actually work though...  
I'll trust you on that.

**Comment by garronej** [comment]
Thank you so much for this work.  
I'm sure the community will greatly appreciate it!

**Comment by garronej** [comment]
Bummer, it dosen't work.  
@lordvlad, can you pick it up from here 
https://github.com/InseeFrLab/keycloakify/pull/233 ?  

here is an invite on for the repo so you can push on the branch:  https://github.com/InseeFrLab/keycloakify/invitations

**Comment by lordvlad** [comment]
@garronej whats broken? it worked for me and github actions were all green?

**Comment by garronej** [comment]
Again @lordvlad,  
I wanted to reiterate and tell you how very sorry I am to call your PR broken when it was I that didn’t run the right command.  
Your effort is very much appreciated!

**Comment by lordvlad** [comment]
haha all good. happy to make an improvement

**Pull request merged**